src/main/java/org/openrefine/extensions/commons/importer/RelatedCategoryFetcher.java

```diff
@@ -0,0 +1,126 @@
+package org.openrefine.extensions.commons.importer;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.refine.expr.EvalError;
+
+import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+
+/*
+ * This class takes an existing iterator over file records, and enriches
+ * the file records with the list of categories they belong to
+ *
+ * @param apiUrl
+ * @param iteratorFileRecords
+ */
+public class RelatedCategoryFetcher implements Iterator<FileRecord> {
+    Iterator<FileRecord> iteratorFileRecords;
+    FileRecord fileRecordNew;
+    int filesIndex;
+    List <FileRecord> fileRecordOriginal;
+    String apiUrl;
+    HttpUrl urlRelatedCategories;
+
+    public RelatedCategoryFetcher(String apiUrl, Iterator<FileRecord> iteratorFileRecords) {
+        this.apiUrl = apiUrl;
+        this.iteratorFileRecords = iteratorFileRecords;
+        fileRecordOriginal = new ArrayList<>();
+    }
+
+    /*
+     * API call for fetching the related categories
+     * @param filename -> change to list of files
+     * @param pageId
+     * @return list of related categories
+     */
+    public List<List<String>> getRelatedCategories(List <FileRecord> fileRecordOriginal) throws IOException {
+
+        String titles = fileRecordOriginal.get(0).fileName;
+        int titlesIndex =1;
+        if (titlesIndex < fileRecordOriginal.size()) {
+            titles += "|" + fileRecordOriginal.get(titlesIndex++).fileName;
+        }
+        OkHttpClient client = new OkHttpClient.Builder().build();
+        urlRelatedCategories = HttpUrl.parse(apiUrl).newBuilder()
+                .addQueryParameter("action", "query")
+                .addQueryParameter("prop", "categories")
+                .addQueryParameter("titles", titles)
+                .addQueryParameter("format", "json").build();
+        Request request = new Request.Builder().url(urlRelatedCategories).build();
+        Response response = client.newCall(request).execute();
+        JsonNode jsonNode = new ObjectMapper().readTree(response.body().string());
+        List<JsonNode> relatedCategories = new ArrayList<>();
+        List<String> toCategoriesColumn = new ArrayList<>();
+        List<List<String>> toCategoriesColumn2 = new ArrayList<>();
+        for (int i = 0; i < fileRecordOriginal.size(); i++) {
+            relatedCategories.add(jsonNode.path("query").path("pages").path(fileRecordOriginal.get(i).pageId).path("categories"));
+            //toCategoriesColumn.add(fileRecordOriginal.get(i).pageId);
+        }
+        for (JsonNode category: relatedCategories) {
+            toCategoriesColumn.add(category.findValue("title").asText());
```

Comment:
Here, `category` is a `JsonNode`, which is actually an array containing multiple categories. So when you do `category.findValue("title")`, it somehow returns the first one it finds, in the first object inside the array.
To debug those sorts of problems, it is really helpful to use your debugger: put a breakpoint on this line, run your test and check if the contents of the variables match your expectations :)